File: src/bmsx/render/host_overlay/pipeline.ts
import type { HostMenuPipelineState, HostOverlayPipelineState } from '../backend/backend';
import { consoleCore } from '../../core/console';
import type { Host2DSubmission } from '../shared/submissions';
import { consumeOverlayFrame, hasPendingOverlayFrame } from './overlay_queue';

const EMPTY_HOST_OVERLAY_COMMANDS: Host2DSubmission[] = [];

export function buildHostOverlayState(): HostOverlayPipelineState {
	const view = consoleCore.view;
	if (hasPendingOverlayFrame()) {
		const frame = consumeOverlayFrame();
		return {
			width: view.offscreenCanvasSize.x,
			height: view.offscreenCanvasSize.y,
			overlayWidth: frame.width,
			overlayHeight: frame.height,
			time: consoleCore.platform.clock.now() / 1000,
			delta: consoleCore.deltatime_seconds,
			commands: frame.commands,
		};
	}
	return {
		width: view.offscreenCanvasSize.x,
		height: view.offscreenCanvasSize.y,
		overlayWidth: view.viewportSize.x,
		overlayHeight: view.viewportSize.y,
		time: consoleCore.platform.clock.now() / 1000,
		delta: consoleCore.deltatime_seconds,
		commands: EMPTY_HOST_OVERLAY_COMMANDS,
	};
}

export function buildHostMenuState(): HostMenuPipelineState {
	const view = consoleCore.view;
	return {
		width: view.offscreenCanvasSize.x,
		height: view.offscreenCanvasSize.y,
		overlayWidth: view.viewportSize.x,
		overlayHeight: view.viewportSize.y,
		time: consoleCore.platform.clock.now() / 1000,
		delta: consoleCore.deltatime_seconds,
	};
}

File: native/machine/render/host_overlay/pipeline.cpp
#include "render/host_overlay/pipeline.h"

#include "core/console.h"
#include "render/gameview.h"
#include "render/host_overlay/overlay_queue.h"

namespace bmsx {
namespace {

void fillHost2DState(Host2DPipelineState& state, GameView& view) {
	state.width = static_cast<i32>(view.offscreenCanvasSize.x);
	state.height = static_cast<i32>(view.offscreenCanvasSize.y);
	state.overlayWidth = static_cast<i32>(view.viewportSize.x);
	state.overlayHeight = static_cast<i32>(view.viewportSize.y);
}

} // namespace

HostOverlayPipelineState buildHostOverlayState() {
	HostOverlayPipelineState state;
	state.time = ConsoleCore::instance().totalTime();
	state.delta = ConsoleCore::instance().deltaTime();
	if (!hasPendingOverlayFrame()) {
		GameView& view = *ConsoleCore::instance().view();
		fillHost2DState(state, view);
		return state;
	}
	const HostOverlayFrame frame = consumeOverlayFrame();
	state.width = frame.renderWidth;
	state.height = frame.renderHeight;
	state.overlayWidth = frame.logicalWidth;
	state.overlayHeight = frame.logicalHeight;
	state.commandKinds = frame.commandKinds;
	state.commandRefs = frame.commandRefs;
	state.commandCount = frame.commandCount;
	return state;
}

HostMenuPipelineState buildHostMenuState(GameView& view) {
	HostMenuPipelineState state;
	fillHost2DState(state, view);
	return state;
}

} // namespace bmsx
